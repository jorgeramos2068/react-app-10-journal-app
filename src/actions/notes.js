import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import loadNotes from '../helpers/loadNotes';
import uploadFile from '../helpers/uploadFile';
import types from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(activateNote(doc.id, newNote));
  };
};

export const activateNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = {...note};
    delete noteToFirestore.id;
    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);
    dispatch(refreshNote(note.id, note));
    Swal.fire('Saved', 'The note was saved successfully', 'success');
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdate,
  payload: {
    id,
    note: {
      id,
      ...note
    }
  }
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const {active: activeNote} = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      }
    });
    const fileUrl = await uploadFile(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));
    Swal.close();
  };
};
