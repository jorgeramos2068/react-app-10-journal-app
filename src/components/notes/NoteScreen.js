import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import useForm from '../../hooks/useForm';
import NotesAppBar from './NotesAppBar';

const NoteScreen = () => {
  const {active:note} = useSelector(state => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);
  const {body, title} = formValues;
  const activeId = useRef(note.id);

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);
  
  const dispatch = useDispatch();

  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="test"
          placeholder="Some awesome title"
          className="notes__title-input"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="What happened today"
          className="notes__textarea"
          value={body}
          onChange={handleInputChange}
        >
        </textarea>
       {
          (note.url) &&
          <div className="notes__image">
            <img
              src="https://www.theguardian.pe.ca/media/photologue/photos/cache/B97946073Z.1_20190906151130_000GMGP7UF1.1-1_large.jpg"
              alt="Notes"
            />
          </div>
       }
      </div>
    </div>
  );
};

export default NoteScreen;
