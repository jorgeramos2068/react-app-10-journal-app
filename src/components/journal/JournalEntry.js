import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activateNote } from '../../actions/notes';

const JournalEntry = ({id, date, title, body, url}) => {
  const noteDate = moment(date);
  const dispatch = useDispatch();

  const handleEntryClick = () => {
    const note = {
      date,
      title,
      body,
      url
    }
    dispatch(activateNote(id, note));
  };

  return (
    <div
      className="journal__entry"
      onClick={handleEntryClick}
    >
      {
        url &&
        <div
          className="journal__entry-picture"
          style={{
            backgroundSize: 'cover',
            backgroundImage: `url(${url})`,
          }}
        >
        </div>
      }
      <div className="journal__entry-body">
        <p className="journal__entry-title">
          {title}
        </p>
        <p className="journal__entry-content">
          {body}
        </p>
      </div>
      <div className="journal__entry-date-box">
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('MMM Do')}</h4>
      </div>
    </div>
  );
};

export default JournalEntry;
