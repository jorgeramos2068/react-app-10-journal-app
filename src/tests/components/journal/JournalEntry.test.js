import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import JournalEntry from '../../../components/journal/JournalEntry';
import { activateNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Tests in JournalEntry component', () => {
  const initialState = {};
  let store;
  let wrapper;
  const note = {
    id: 'test_id',
    title: 'Test title',
    body: 'Test body',
    date: 0,
    url: 'https://test.com/test.jpg'
  }

  beforeEach(() => {
    jest.clearAllMocks();
    store = mockStore(initialState);
    store.dispatch = jest.fn();
    wrapper = mount(
      <Provider store={store}>
        <JournalEntry {...note} />
      </Provider>
    );
  });

  test('Should display JournalEntry correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should active the current note', () => {
    wrapper.find('.journal__entry').prop('onClick')();
    expect(store.dispatch).toHaveBeenCalledWith(
      activateNote(note.id, {...note})
    );
  });
});
