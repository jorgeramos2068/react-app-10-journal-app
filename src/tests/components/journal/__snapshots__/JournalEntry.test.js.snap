// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in JournalEntry component Should display JournalEntry correctly 1`] = `
<div
  className="journal__entry animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://test.com/test.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Test title
    </p>
    <p
      className="journal__entry-content"
    >
      Test body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      Dec 31st
    </h4>
  </div>
</div>
`;
